New topic - gRPC interceptors on .NET (#25150)
dotnet/AspNetCore.Docs: aspnetcore/grpc/interceptors.md

---
title: gRPC interceptors on .NET

> author: erni27

Reviewer: ```suggestion
author: erni27
ms.author: jamesnk
```
Adding `ms.author: jamesnk` will fix the build failure.

var invoker = channel.Intercept(new ClientLoggerInterceptor());

var client = new Greeter.GreeterClient(invoker);
```

The `Intercept(...)` extension method can be chained together to configure multiple interceptors for a channel. Alternatively, there is an `Intercept` overload that accepts multiple interceptors. Any number of interceptors can be executed for a single gRPC call.

```csharp
var invoker = channel
    .Intercept(new ClientTokenInterceptor())
    .Intercept(new ClientMonitoringInterceptor())
    .Intercept(new ClientLoggerInterceptor());
```

The order of the parameters matters so in the preceding code, interceptors will be invoked as follow `ClientLoggerInterceptor`, `ClientMonitoringInterceptor` and then `ClientTokenInterceptor`.

For information on how to configure interceptors with gRPC client factory, see [Configure Interceptors](xref:grpc/clientfactory#configure-interceptors).

## Server interceptors

gRPC server interceptors intercept incoming RPC requests. They provide access to the incoming request, the outgoing response and the context for a server-side call.

`Interceptor` methods to override for server:

> * `UnaryServerHandler` - intercepts an incoming unary RPC.

Reviewer 2: ```suggestion
* `UnaryServerHandler` - intercepts a unary RPC.
```

* `ClientStreamingServerHandler` - intercepts a client streaming RPC.
* `ServerStreamingServerHandler` - intercepts a server streaming RPC.
* `DuplexStreamingServerHandler` - intercepts a bidirectional streaming RPC.

### Create a server gRPC interceptor

The following code presents an example of an intercepting an incoming unary RPC.

```csharp
public class ServerLoggingInterceptor : Interceptor
{
    private readonly ILogger _logger;

    public ServerLoggingInterceptor(ILogger<ServerLoggingInterceptor> logger)
    {
        _logger = logger;
    }

    public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
        TRequest request,
        ServerCallContext context,
        UnaryServerMethod<TRequest, TResponse> continuation)
    {

>         _logger.LogWarning($"Starting receiving call. Type: {MethodType.Unary}. Method: {context.Method}. Request: {typeof(TRequest)}. Response: {typeof(TResponse)}");

Reviewer 2: ```suggestion
        _logger.LogInformation($"Starting receiving call. Type: {MethodType.Unary}. Method: {context.Method}.");
```

        try
        {
            return await continuation(request, context);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error thrown by {context.Method}.");
            throw;
        }
    }
}
```

> Overridden `UnaryServerHandler` does the following:
> 
> * Intercepts an incoming unary call.
> * Logs the call.
> * Logs an exception if occured.

Reviewer 2: ```suggestion
Overridden `UnaryServerHandler` does the following:

* Intercepts an incoming unary call.
* Logs details about the call.
* Calls the `continuation` parameter passed into the method. This invokes the next interceptor in the chain or the service handler if this is the last interceptor.
* Logs an exception if occurred.
```

* `AsyncClientStreamingCall` - intercepts an asynchronous invocation of a client streaming RPC.
* `AsyncServerStreamingCall` - intercepts an asynchronous invocation of a server streaming RPC.
* `AsyncDuplexStreamingCall` - intercepts an asynchronous invocation of a bidirectional streaming RPC.

> [!WARNING]
> Although both `BlockingUnaryCall` and `AsyncUnaryCall` refer to unary RPCs, they can't be used interchangeably. A blocking invocation would not be intercepted by `AsyncUnaryCall` and an asynchronous one by `BlockingUnaryCall`.

### Create a client gRPC interceptor

The following code presents an example of intercepting an asynchronous invocation of a simple remote call.

```csharp
public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
    TRequest request,
    ClientInterceptorContext<TRequest, TResponse> context,
    AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
{
    LogCall(context.Method);
    AddCallerMetadata(ref context);

    var call = continuation(request, context);

    return new AsyncUnaryCall<TResponse>(HandleResponse(call.ResponseAsync), call.ResponseHeadersAsync, call.GetStatus, call.GetTrailers, call.Dispose);
}

> ```

Reviewer 2: This is a pretty complex interceptor (adding metadata by replacing the context, handling async response), and it has a lot of methods that aren't shown. I think the most understandable example would:
- include the top-level Interceptor being inherited from.
- only show logging the call

Something like:
```cs
public class ClientLoggingInterceptor : Interceptor
{
    private readonly ILogger _logger;

    public LoggingInterceptor(ILoggerFactory loggerFactory)
    {
        _logger = loggerFactory.CreateLogger<LoggingInterceptor>();
    }

    public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
        TRequest request,
        ClientInterceptorContext<TRequest, TResponse> context,
        AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
    {
       _logger.LogInformation($"Starting call. Type: {method.Type}. Request: {typeof(TRequest)}. Response: {typeof(TResponse)}");
        return continuation(request, context);
    }
}
```
I think a more complex example can be added later. Using async is a common question and the more complex example can discuss how to via overriding `ResponseAsync`

```csharp
private void AddCallerMetadata<TRequest, TResponse>(ref ClientInterceptorContext<TRequest, TResponse> context)
    where TRequest : class
    where TResponse : class
{
    var headers = context.Options.Headers;

    // Call doesn't have a headers collection to add to.
    // Need to create a new context with headers for the call.
    if (headers == null)
    {
        headers = new Metadata();
        var options = context.Options.WithHeaders(headers);
        context = new ClientInterceptorContext<TRequest, TResponse>(context.Method, context.Host, options);
    }

    // Add caller metadata to call headers
    headers.Add("caller-user", Environment.UserName);
    headers.Add("caller-machine", Environment.MachineName);
    headers.Add("caller-os", Environment.OSVersion.ToString());
}
```

`LogCall`, `AddCallerMetadata` and `HandleResponse` methods are private methods embedded in the exemplary `ClientLoggerInterceptor` class. For complete implementation, see an [example of client interceptor](https://github.com/grpc/grpc-dotnet/blob/master/examples/Interceptor/Client/ClientLoggerInterceptor.cs).

### Configure client interceptors

> gRPC client interceptors are configured by creating a call invoker from a channel and then 
> creating a client from that invoker.
> 
> The following code:
> * Creates a channel by using `GrpcChannel.ForAddress`.
> * Creates an invoker by using `Intercept` method on the created channel.
> * Creates a client from the invoker.

Reviewer 2: ```suggestion
gRPC client interceptors are configured on a channel.

The following code:
* Creates a channel by using `GrpcChannel.ForAddress`.
* Uses the `Intercept(...)` extension method to configure the channel to use the interceptor. Note that this method returns a `CallInvoker`. Strongly typed gRPC clients may be created from an invoker just like a channel.
* Creates a client from the invoker. gRPC calls made by the client automatically execute the interceptor.
```
IMO because a client can be created from a channel and call invoker in the same way there isn't a need to stress their differences.

```csharp
using var channel = GrpcChannel.ForAddress("https://localhost:5001");
var invoker = channel.Intercept(new ClientLoggerInterceptor());

var client = new Greeter.GreeterClient(invoker);
```

> A call invoker intercepts the channel with the given interceptor. If you want to intercept the channel with multiple interceptors use overloaded `Intercept` method. Any number of interceptors may be used for a single RPC.

Reviewer 2: ```suggestion
The `Intercept(...)` extension method can be chained together to configure multiple interceptors for a channel. Alternatively, there is an `Intercept` overload that accepts multiple interceptors. Any number of interceptors can be executed for a single gRPC call.
```

> ```csharp
> var invoker = channel.Intercept(
>     new ClientLoggerInterceptor(),
>     new ClientMonitoringInterceptor(),
>     new ClientTokenInterceptor());
> ```

Reviewer 2: ```suggestion
```csharp
var invoker = channel
    .Intercept(new ClientTokenInterceptor())
    .Intercept(new ClientMonitoringInterceptor())
    .Intercept(new ClientLoggerInterceptor());
```
```
IMO focus on chaining calls and mention there is a helper method for combining them

The order of the parameters matters so in the preceding code, interceptors will be invoked as follow `ClientLoggerInterceptor`, `ClientMonitoringInterceptor` and then `ClientTokenInterceptor`.

> The same result can be achieved by building a chain of interceptors.
> 
> ```csharp
> var invoker = channel
>     .Intercept(new ClientTokenInterceptor())
>     .Intercept(new ClientMonitoringInterceptor())
>     .Intercept(new ClientLoggerInterceptor());
> ```
> 

Reviewer 2: ```suggestion
```
IMO I think this can be removed for brevity.

> For information on how to configure interceptors with gRPC client factory, see [Configure Channel and Interceptors](xref:grpc/clientfactory#configure-channel-and-interceptors).

Reviewer 2: ```suggestion
For information on how to configure interceptors with gRPC client factory, see [Configure Interceptors](xref:grpc/clientfactory#configure-interceptors).
```

## Server interceptors

gRPC server interceptors intercept incoming RPC requests. They provide access to the incoming request, the outgoing response and the context for a server-side call.

`Interceptor` methods to override for server:
* `UnaryServerHandler` - intercepts an incoming unary RPC.
* `ClientStreamingServerHandler` - intercepts a client streaming RPC.
* `ServerStreamingServerHandler` - intercepts a server streaming RPC.
* `DuplexStreamingServerHandler` - intercepts a bidirectional streaming RPC.

### Create a server gRPC interceptor

The following code presents an example of an intercepting an incoming unary RPC.

```csharp
public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
    TRequest request,
    ServerCallContext context,
    UnaryServerMethod<TRequest, TResponse> continuation)
{
    LogCall<TRequest, TResponse>(MethodType.Unary, context);

    try
    {
        return await continuation(request, context);
    }
    catch (Exception ex)
    {
        // Note: The gRPC framework also logs exceptions thrown by handlers to .NET Core logging.
        _logger.LogError(ex, $"Error thrown by {context.Method}.");

        throw;
    }
}

> ```

Reviewer 2: Make changes similar to the client interceptor to simplify

* `context` carries metadata associated with the server-side call like request metadata, deadlines and cancellation or RPC result.

The following code shows how to retrieve request metadata from the context.

```csharp
private void LogCall<TRequest, TResponse>(MethodType methodType, ServerCallContext context)
    where TRequest : class
    where TResponse : class
{
    _logger.LogWarning($"Starting call. Type: {methodType}. Request: {typeof(TRequest)}. Response: {typeof(TResponse)}");
    WriteMetadata(context.RequestHeaders, "caller-user");
    WriteMetadata(context.RequestHeaders, "caller-machine");
    WriteMetadata(context.RequestHeaders, "caller-os");

    void WriteMetadata(Metadata headers, string key)
    {
        var headerValue = headers.SingleOrDefault(h => h.Key == key)?.Value;
        _logger.LogWarning($"{key}: {headerValue ?? "(unknown)"}");
    }
}
```

For complete implementation of introduced `ServerLoggerInterceptor`, see an [example of server interceptor](https://github.com/grpc/grpc-dotnet/blob/master/examples/Interceptor/Server/ServerLoggerInterceptor.cs).

### Configure server interceptors

> gRPC server interceptors can be configured for all services by providing an appropriate options delegate to the `AddGrpc` call in `Startup.ConfigureServices`.

Reviewer 2: ```suggestion
gRPC server interceptors are configured at startup. The following code:

* Adds gRPC to the app with `AddGrpc`.
* Configures `ServerLoggerInterceptor` for all services by adding it to the service option's `Interceptors` collection.
```

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddGrpc(options =>
    {
        options.Interceptors.Add<ServerLoggerInterceptor>();
    });
}
```

> If a particular interceptor is dedicated for one specific service, then configure options for a single service.

Reviewer 2: ```suggestion
An interceptor can also be configured for a specific service by using `AddServiceOptions` and specifying the service type.
```
